transformer_engine/jax/flax/transformer.py

```diff
@@ -248,15 +251,51 @@ def rotary_pos_emb(x: Array, windows: Tuple[int, int], transpose_batch_sequence:
     positions = jnp.expand_dims(jnp.arange(x.shape[seq_dim]), axis=batch_dim)
     positions = jnp.expand_dims(positions, axis=tuple(range(2, x.ndim)))
 
-    sinusoidal_positions = positions / time_scales
-    sin = jnp.sin(sinusoidal_positions)
-    cos = jnp.cos(sinusoidal_positions)
+    def generate_sin_cos(timescales):
+        sinusoidal_positions = positions / timescales
+        sin = jnp.sin(sinusoidal_positions)
+        cos = jnp.cos(sinusoidal_positions)
+        return sin, cos
+
+    def alternate_impl():
+        sin, cos = generate_sin_cos(time_scales)
+
+        x1, x2 = jnp.split(x, 2, axis=-1)
+        part_1 = (x1 * cos - x2 * sin).astype(x.dtype)
+        part_2 = (x2 * cos + x1 * sin).astype(x.dtype)
+
+        output = jnp.concatenate([part_1, part_2], axis=-1)
+        return output
+
+    def consecutive_impl():
+        sin, cos = generate_sin_cos(jnp.repeat(time_scales, 2, axis=-1))
+
+        x_shifted_left = jnp.roll(x, -1, axis=-1)
+        x_shifted_right = jnp.roll(x, 1, axis=-1)
+        x_shifted = jax.lax.select(
+            jnp.tile(
+                jnp.mod(jnp.arange(embed_dim, dtype=jnp.int32), 2),
+                x.shape[:-1] + (1,),
+            ),
+            x_shifted_right,
+            x_shifted_left,
+        )
+
+        sign = jnp.sign(jnp.mod(jnp.arange(embed_dim, dtype=jnp.int32), 2) - 0.5)
+
+        output = x * cos + x_shifted * sin * sign
+        return output
+
+    def canonicalize_group_method(gm):
+        canonicalized_gm = gm.lower().strip().replace('-', '').replace('_', '')
+        assert canonicalized_gm in ['consecutive', 'alternate']
```

Comment:
This probably needs a helpful error message like `Invalid relative positional embedding method: must be either 'alternate' or 'consecutive'.`

Reply:
Added.